fs/fs.go

```diff
@@ -0,0 +1,299 @@
+	}
+	id, err := strconv.ParseUint(b[i+1:], 10, 64)
+	if err != nil {
+		return 0, xerrors.Errorf("parsing sector id (name: '%s'): %w", b, err)
+	}
+
+	return id, nil
+}
+
+func (p SectorPath) miner() (address.Address, error) {
+	b := filepath.Base(string(p))
+	fi := strings.IndexByte(b, '-')
+	li := strings.LastIndexByte(b, '-')
+	if li < 0 || fi < 0 {
+		return address.Undef, xerrors.Errorf("malformed sector file name: '%s', expected to be in form 's-[miner]-[id]'", b)
+	}
+
+	return address.NewFromString(b[fi+1 : li])
+}
+
+func SectorName(miner address.Address, sectorID uint64) string {
+	return fmt.Sprintf("s-%s-%d", miner, sectorID)
+}
+
+func (p StoragePath) Sector(typ DataType, miner address.Address, id uint64) SectorPath {
+	return SectorPath(filepath.Join(string(p), string(typ), SectorName(miner, id)))
+}
+
+type pathInfo struct {
+	cache  bool // TODO: better name?
+	weight int
+}
+
+type FS struct {
+	paths map[StoragePath]*pathInfo
+
+	// in progress actions
+
+	// path -> datatype
+	reserved map[StoragePath]map[DataType]uint64
+
+	locks map[SectorPath]chan struct{}
+
+	lk sync.Mutex
+}
+
+type PathConfig struct {
+	Path string
+
+	Cache  bool
+	Weight int
+}
+
+func OpenFs(cfg []PathConfig) *FS {
+	paths := map[StoragePath]*pathInfo{}
+	for _, c := range cfg {
+		paths[StoragePath(c.Path)] = &pathInfo{
+			cache:  c.Cache,
+			weight: c.Weight,
+		}
+	}
+	return &FS{
+		paths:    paths,
+		reserved: map[StoragePath]map[DataType]uint64{},
+		locks:    map[SectorPath]chan struct{}{},
+	}
+}
+
+func (f *FS) Init() error {
+	for path := range f.paths {
+		for _, dir := range []string{string(path),
+			filepath.Join(string(path), string(DataCache)),
+			filepath.Join(string(path), string(DataStaging)),
+			filepath.Join(string(path), string(DataSealed)),
+			filepath.Join(string(path), string(DataUnsealed))} {
+			if err := os.Mkdir(dir, 0755); err != nil {
+				if os.IsExist(err) {
+					continue
+				}
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// TODO: fix the locking situation
+
+func (f *FS) FindSector(typ DataType, miner address.Address, id uint64) (out SectorPath, err error) {
+	// TODO: consider keeping some sort of index at some point
+
+	for path := range f.paths {
+		p := path.Sector(typ, miner, id)
+
+		_, err := os.Stat(string(p))
+		if os.IsNotExist(err) {
+			continue
+		}
+		if err != nil {
+			log.Errorf("error scanning path %s for sector %s (%s): %+v", p, SectorName(miner, id), string(typ))
+			continue
+		}
+		if out != "" {
+			if !f.paths[p.storage()].cache {
+				log.Errorf("%s also found in cache at %s", p, out)
+				return p, nil
+			}
+		}
+		out = p
+	}
+
+	if out == "" {
+		return "", ErrNotFound
+	}
+
+	return out, nil
+}
+
+func (f *FS) findBestPath(size uint64, cache bool, strict bool) StoragePath {
+	var best StoragePath
+	bestw := big.NewInt(0)
+
+	// If we need cache, only return cache. If we need storage, prefer storage, fall back to cache
```

Comment:
I would add a special error for the case where we need cache, but have no available cache space